lib/client.js

```diff
@@ -1,19 +1,85 @@
-var version      = require('version'),
-    Promise      = require('../vendor/ayepromise'),
-    requestCount = 0;
+var PROMISE_TIMEOUT = 5000, // 5 seconds
+    ZAF_EVENT       = /^zaf\./,
+    version         = require('version'),
+    Promise         = window.Promise || require('../vendor/ayepromise'),
+    pendingPromises = {},
+    promiseCount    = 0,
+    requestCount    = 0;
+
+function rawPostMessage(client, msg, forceReady) {
+  if (client.ready || forceReady) {
+    client._source.postMessage(msg, client._origin);
+  } else {
+    client.on('app.registered', rawPostMessage.bind(null, client, msg));
+  }
+}
+
+function removePromise(id) {
+  delete pendingPromises[id];
+}
+
+function isValidEvent(client, event) {
+  return client._origin === event.origin && client._source === event.source;
+}
+
+function messageHandler(client, event) {
+  if (!isValidEvent(client, event)) { return; }
+
+  var data = event.data;
+
+  if (!data) { return; }
+
+  if (typeof data === 'string') {
+    try {
+      data = JSON.parse(event.data);
+    } catch (e) {
+      return;
+    }
+  }
+
+  console.debug('SDK <<', data);
+
+  var pendingPromise;
+  if (data.id && (pendingPromise = pendingPromises[data.id])) {
+    if (data.error) {
+      var err = data.error;
+      if (err.code) {
+        err = new Error(data.error.msg);
+        err.name = data.error.code;
+        err.stack = data.error.stack;
+      }
+      pendingPromise.reject(err);
+    } else {
+      pendingPromise.resolve(data.result);
+    }
+  } else if (ZAF_EVENT.test(data.key)) {
+    var key = data.key.replace(ZAF_EVENT, '');
+    client && client.trigger(key, data.message);
```

Comment:
@sandlerr you want to remove events? or just restructure the message format for them?